File: app/src/main/java/com/theappsolutions/boilerplate/util/storage_utils/RealmUtils.java
package com.theappsolutions.boilerplate.util.storage_utils;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmModel;

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 *
 * Base Realm operations wrapper
 */
public class RealmUtils {

    public static <E extends RealmModel> E copyFromRealm(E obj, Realm realm) {
        if (obj == null) return null;
        return realm.copyFromRealm(obj);
    }

    public static <E extends RealmModel> List<E> copyFromRealm(
            Iterable<E> obj, Realm realm) {
        if (obj == null) return null;
        return realm.copyFromRealm(obj);
    }

    public static <E extends RealmModel> void saveToRealm(E obj) {
        Realm realm = getRealm();
        realm.executeTransaction(realm1 -> realm1.copyToRealmOrUpdate(obj));
        realm.close();
    }

    public static <E extends RealmModel> void saveToRealm(List<E> list) {
        Realm realm = getRealm();
        realm.executeTransaction(realm1 -> realm1.copyToRealmOrUpdate(list));
        realm.close();
    }

    public static <E extends RealmModel> void onlySaveToRealm(E obj) {
        Realm realm = getRealm();
        realm.executeTransaction(realm1 -> realm1.copyToRealm(obj));
        realm.close();
    }

    public static void clearRealmTable(Class<? extends RealmModel> realmClass) {
        Realm realm = getRealm();
        realm.executeTransaction(realm1 -> realm1.delete(realmClass));
        realm.close();
    }

    public static Realm getRealm() {
        return Realm.getDefaultInstance();
    }
}

File: app/src/main/java/com/theappsolutions/boilerplate/util/storage/RealmUtils.kt
package com.theappsolutions.boilerplate.util.storage

import io.realm.Realm
import io.realm.RealmModel

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 *
 * Base Realm operations wrapper
 */
object RealmUtils {

    fun getRealm(): Realm = Realm.getDefaultInstance()

    fun <E : RealmModel> copyFromRealm(obj: E?, realm: Realm): E? {
        return when (obj) {
            null -> null
            else -> realm.copyFromRealm(obj)
        }
    }

    fun <E : RealmModel> copyFromRealm(
            obj: Iterable<E>?, realm: Realm): List<E>? {
        return when (obj) {
            null -> null
            else -> realm.copyFromRealm(obj)
        }
    }

    fun <E : RealmModel> saveToRealm(obj: E) {
        val realm = getRealm()
        realm.executeTransaction { it.copyToRealmOrUpdate(obj) }
        realm.close()
    }

    fun <E : RealmModel> saveToRealm(list: List<E>) {
        val realm = getRealm()
        realm.executeTransaction { it.copyToRealmOrUpdate(list) }
        realm.close()
    }

    fun <E : RealmModel> onlySaveToRealm(obj: E) {
        val realm = getRealm()
        realm.executeTransaction { it.copyToRealm(obj) }
        realm.close()
    }

    fun clearRealmTable(realmClass: Class<out RealmModel>) {
        val realm = getRealm()
        realm.executeTransaction { realm1 -> realm1.delete(realmClass) }
        realm.close()
    }
}
